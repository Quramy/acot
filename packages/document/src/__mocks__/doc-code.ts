import type { PartialDeep } from 'type-fest';
import type { DocCode } from '../doc-code';
import { merge } from './merge';

export const createDocCode = (values: PartialDeep<DocCode> = {}): DocCode =>
  merge<DocCode>(
    {
      path: '',
      rule: '',
      summary: {
        markdown: '',
        text: '',
      },
      type: 'correct',
      id: '',
      meta: {},
      html: '',
    },
    values,
  );
