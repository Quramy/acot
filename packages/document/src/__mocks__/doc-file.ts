import type { DocFile } from '../doc-file';

export const createDocFile = (partial: Partial<DocFile> = {}): DocFile => ({
  path: '',
  content: '',
  ...partial,
});
