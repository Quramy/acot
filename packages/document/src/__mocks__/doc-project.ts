import type { DocProject } from '../doc-project';

export const createDocProject = (
  partial: Partial<DocProject> = {},
): DocProject => ({
  root: '',
  name: '',
  main: '',
  preset: {
    id: '',
    rules: new Map(),
    configs: new Map(),
  },
  codes: [],
  ...partial,
});
