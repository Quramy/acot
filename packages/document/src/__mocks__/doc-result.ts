import type { DocResult } from '../doc-result';

export const createDocResult = (
  partial: Partial<DocResult> = {},
): DocResult => ({
  skips: [],
  passes: [],
  errors: [],
  ...partial,
});
