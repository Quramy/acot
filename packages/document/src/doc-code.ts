export type DocCodeType = 'correct' | 'incorrect';

export type DocCodeMeta = {
  'acot-ignore'?: boolean;
  'acot-head'?: boolean;
  'acot-template'?: string;
  [key: string]: string | number | boolean | null | undefined;
};

export type DocCodeSummary = {
  markdown: string;
  text: string;
};

export type DocCode = {
  path: string;
  rule: string;
  summary: DocCodeSummary;
  type: DocCodeType;
  id: string;
  meta: DocCodeMeta;
  html: string;
};

const reservedMetaKeys = ['acot-ignore', 'acot-head', 'acot-template'] as const;

export const extractCodeMeta = (
  code: DocCode,
): Omit<DocCodeMeta, typeof reservedMetaKeys[number]> => {
  return Object.entries(code.meta).reduce<DocCodeMeta>((acc, [key, value]) => {
    if (!reservedMetaKeys.includes(key as any)) {
      acc[key] = value;
    }
    return acc;
  }, {});
};

export const generateDocPath = ({ rule, id }: DocCode): string =>
  `/${rule}/${id}`;

export const generateDocUrl = (port: number, code: DocCode): string =>
  `http://localhost:${port}${generateDocPath(code)}`;
