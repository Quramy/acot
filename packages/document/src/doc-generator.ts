import path from 'path';
import { pkg2shorthand } from '@acot/utils';
import type { DocCode } from './doc-code';
import type { DocProject } from './doc-project';

export class DocGenerator {
  public async generate(project: DocProject): Promise<string> {
    const map = new Map<string, DocCode>();
    project.codes.forEach((code) => {
      map.set(code.path, code);
    });

    if (map.size === 0) {
      return '';
    }

    const codes = Array.from(map.values());
    const table = [['Name', 'Summary', ':heavy_check_mark:']];
    const prefix = pkg2shorthand(project.name, 'preset');

    for (const code of codes) {
      const ruleId = `${project.preset.id}/${code.rule}`;
      const rule = project.preset.rules.get(ruleId);
      const relative = path.relative(project.root, code.path);

      table.push([
        `[\`${prefix}/${code.rule}\`](./${relative})`,
        code.summary.markdown,
        rule?.meta?.recommended ? ':heavy_check_mark:' : '',
      ]);
    }

    const max = table.reduce((acc, cur) => {
      return acc.map((v, i) => Math.max(v, cur[i].length));
    }, Array(table[0].length).fill(0));

    return table
      .reduce((acc, cur, i) => {
        acc.push(this._row(cur.map((col, j) => col.padEnd(max[j]))));

        if (i === 0) {
          acc.push(this._row(max.map((m) => `:${'-'.repeat(m - 1)}`)));
        }

        return acc;
      }, [])
      .join('\n');
  }

  private _row(col: string[]): string {
    return `| ${col.join(' | ')} |`;
  }
}
