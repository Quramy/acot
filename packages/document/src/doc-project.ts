import type { Preset } from '@acot/types';
import type { DocCode } from './doc-code';

export type DocProject = {
  root: string;
  name: string;
  main: string;
  preset: Preset;
  codes: DocCode[];
};
