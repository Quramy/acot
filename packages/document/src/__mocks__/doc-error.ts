import type { DocError } from '../doc-error';
import { createDocCode } from './doc-code';

export const createDocError = (partial: Partial<DocError> = {}): DocError => ({
  message: '',
  code: createDocCode(),
  results: [],
  ...partial,
});
