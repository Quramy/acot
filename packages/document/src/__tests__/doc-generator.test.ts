import { createRule } from '@acot/core';
import type { Rule } from '@acot/types';
import { DocGenerator } from '../doc-generator';
import { createDocProject } from '../__mocks__/doc-project';
import { createDocCode } from '../__mocks__/doc-code';

describe('DocGenerator', () => {
  let generator: DocGenerator;

  const factory = (description: string, recommended: boolean): Rule =>
    createRule({
      type: 'global',
      meta: {
        description,
        recommended,
      },
      test: async () => {},
    });

  beforeEach(() => {
    generator = new DocGenerator();
  });

  test('basic', async () => {
    const output = await generator.generate(
      createDocProject({
        root: '/root',
        name: 'acot-plugin-test',
        plugin: {
          id: 'test',
          rules: new Map([
            ['test/rule1', factory('rule1 desc', false)],
            ['test/rule2', factory('rule2 desc', true)],
            [
              'test/rule3-long-rule-identity',
              factory('test/rule3-long-rule-identity desc', true),
            ],
          ]),
          configs: new Map(),
        },
        codes: [
          createDocCode({
            path: '/root/docs/rules/rule1.md',
            rule: 'rule1',
          }),
          createDocCode({
            path: '/root/docs/rules/rule2.md',
            rule: 'rule2',
          }),
          createDocCode({
            path: '/root/docs/rules/rule3-long-rule-identity.md',
            rule: 'rule3-long-rule-identity',
          }),
        ],
      }),
    );

    expect(output).toMatchSnapshot();
  });

  test('scoped package', async () => {
    const output = await generator.generate(
      createDocProject({
        root: '/root',
        name: '@scope/acot-plugin',
        plugin: {
          id: '@scope',
          rules: new Map([
            ['@scope/rule1', factory('rule1 desc', false)],
            ['@scope/rule2', factory('rule2 desc', true)],
          ]),
          configs: new Map(),
        },
        codes: [
          createDocCode({
            path: '/root/docs/rules/rule1.md',
            rule: 'rule1',
          }),
          createDocCode({
            path: '/root/docs/rules/rule2.md',
            rule: 'rule2',
          }),
        ],
      }),
    );

    expect(output).toMatchSnapshot();
  });

  test('empty', async () => {
    const output = await generator.generate(
      createDocProject({
        root: '/root',
        name: 'acot-plugin-test',
        plugin: {
          id: 'test',
          rules: new Map(),
          configs: new Map(),
        },
        codes: [],
      }),
    );

    expect(output).toBe('');
  });
});
