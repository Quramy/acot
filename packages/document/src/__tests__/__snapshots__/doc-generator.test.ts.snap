// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocGenerator basic 1`] = `
"| Name                                                                        | Summary                            | :heavy_check_mark: |
| :-------------------------------------------------------------------------- | :--------------------------------- | :----------------- |
| [\`test/rule1\`](./docs/rules/rule1.md)                                       | rule1 desc                         |                    |
| [\`test/rule2\`](./docs/rules/rule2.md)                                       | rule2 desc                         | :heavy_check_mark: |
| [\`test/rule3-long-rule-identity\`](./docs/rules/rule3-long-rule-identity.md) | test/rule3-long-rule-identity desc | :heavy_check_mark: |"
`;

exports[`DocGenerator scoped package 1`] = `
"| Name                                    | Summary    | :heavy_check_mark: |
| :-------------------------------------- | :--------- | :----------------- |
| [\`@scope/rule1\`](./docs/rules/rule1.md) | rule1 desc |                    |
| [\`@scope/rule2\`](./docs/rules/rule2.md) | rule2 desc | :heavy_check_mark: |"
`;
